pubsub/stanpubsub/stan.go

```diff
@@ -0,0 +1,538 @@
+// The subject is the NATS Subject to subscribe to;
+// for more info, see https://nats.io/documentation/writing_applications/subjects.
+func OpenSubscription(nc stan.Conn, subject, queueGroup string, opts ...stan.SubscriptionOption) (*pubsub.Subscription, error) {
+	ds, maxInflight, err := openSubscription(nc, subject, queueGroup, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewSubscription(ds, &batcher.Options{MaxBatchSize: maxInflight}, nil), nil
+}
+
+// openTopic returns the driver for OpenTopic. This function exists so the test
+// harness can get the driver interface implementation if it needs to.
+func openTopic(nc stan.Conn, subject string) (driver.Topic, error) {
+	if nc == nil {
+		return nil, errors.New("stanpubsub: stan.Conn is required")
+	}
+	return &topic{nc, subject}, nil
+}
+
+type topic struct {
+	nc      stan.Conn
+	subject string
+}
+
+// SendBatch implements driver.Topic interface.
+func (t *topic) SendBatch(ctx context.Context, ms []*driver.Message) error {
+	if t == nil || t.nc == nil {
+		return errNotInitialized
+	}
+
+	var buf bytes.Buffer
+	for _, m := range ms {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		payload, err := encodeMessage(&buf, m)
+		if err != nil {
+			return err
+		}
+		buf.Reset()
+
+		if m.BeforeSend != nil {
+			asFunc := func(i interface{}) bool { return false }
+			if err := m.BeforeSend(asFunc); err != nil {
+				return err
+			}
+		}
+		if err := t.nc.Publish(t.subject, payload); err != nil {
+			return err
+		}
+
+		if m.AfterSend != nil {
+			asFunc := func(i interface{}) bool { return false }
+			if err := m.AfterSend(asFunc); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// IsRetryable implements driver.Topic interface.
+func (t *topic) IsRetryable(err error) bool { return false }
+
+// As implements driver.Topic interface.
+func (t *topic) As(i interface{}) bool {
+	nc, ok := i.(*stan.Conn)
+	if !ok {
+		return false
+	}
+	*nc = t.nc
+	return true
+}
+
+// ErrorAs implement driver.Topic interface.
+func (t *topic) ErrorAs(_ error, _ interface{}) bool { return false }
+
+// ErrorCode implements driver.Topic interface.
+func (t *topic) ErrorCode(err error) gcerrors.ErrorCode {
+	switch err {
+	case nil:
+		return gcerrors.OK
+	case context.Canceled:
+		return gcerrors.Canceled
+	case errNotInitialized:
+		return gcerrors.NotFound
+	case nats.ErrBadSubject, stan.ErrBadConnection:
+		return gcerrors.FailedPrecondition
+	case nats.ErrAuthorization:
+		return gcerrors.PermissionDenied
+	case nats.ErrMaxPayload, nats.ErrReconnectBufExceeded:
+		return gcerrors.ResourceExhausted
+	case stan.ErrTimeout:
+		return gcerrors.DeadlineExceeded
+	}
+	return gcerrors.Unknown
+}
+
+// Close implements driver.Topic interface.
+func (t *topic) Close() error { return nil }
+
+func openSubscription(nc stan.Conn, subject, queueGroup string, options ...stan.SubscriptionOption) (_ driver.Subscription, maxInflight int, err error) {
```

Comment:
It is just called `queue` above, please be consistent.

Reply:
Changed all of them to `queue`. Thanks.